### DISCUSSION TITLE
Official tokio-postgres example w/ bb8 is slow

### DISCUSSION THREAD

**Discussion by greg0x** [tokio-rs/axum#2493] [Q&A] [answered]
### Summary

Hey,

I have been benchmarking and found that the tokio-postgres using bb8 for connection pooling is slow compared to db connection sharing through an Arc.

https://github.com/tokio-rs/axum/tree/main/examples/tokio-postgres - official example 
https://github.com/TechEmpower/FrameworkBenchmarks/blob/master/frameworks/Rust/axum/src/main_pg.rs - tech empower implementation 
https://github.com/greg-nagy/framework-test-at-home - the tests I ran

Do you have any tips on why this is happening? 
Or did I make a mistake somewhere?

### axum version

0.7.3

**Comment by davidpdrsn** [comment, marked as answer]
This is not an axum issue. I’d recommend you ask a more general place.

**Reply by greg0x** [reply]
Do you think this approach could be used in the tokio-postgres example?

I was hoping if this approach turns out to be correct that could be improved and provide a more performant solution.

**Reply by davidpdrsn** [reply]
No. I still think the recommended approach is to use a connection pool.

**Reply by greg0x** [reply]
hm, I probably don't understand the problem well enough to understand why is this the recommended approach. 

thanks for the response!
I will keep digging and see where I land.

**Reply by davidpdrsn** [reply]
Wouldn’t you at least want one connection per core?

**Reply by greg0x** [reply]
I thought I had that when I wrote this question, for sure. I had the wrong mental model about how tokio works. 
Your question helped me understand this. Thank you. 

For my current use case — the counter implementation — using the shared connection is fine. (But that's hardly a generic use case)
I agree with you that using a connection pool is the best practice for general purposes. 

My problem with bb8 is that it causes a ~3x slowdown, and as a result, the performance of the rust solution is ~75% compared to nodejs or go. (This is especially annoying because in those implementations I use a connection pool.)

But this is the current state of things, and it's not an axum issue for sure.

If you have any thoughts on this topic, I would appreciate them, but I also understand that this is not an axum issue.

**Reply by jplatte** [reply]
Maybe try with deadpool instead of bb8? If it also supports tokio-postgres, that is.
